### DISCUSSION TITLE
Interest in Google Summer of Code 2024

### DISCUSSION THREAD

**Discussion by imperorrp** [arviz-devs/arviz#2335] [General]
Hi! I'm interested in applying for one of the ArviZ [GSoC 2024 projects](https://github.com/arviz-devs/arviz/wiki/GSoC-2024-projects), although I haven't decided which one exactly- either the 1. ArviZ Dashboards or 2. Plotting Restructuring projects. I have been looking through and getting acquainted with ArviZ and have submitted 2 PRs (both just very slight bugfixes) so far. 

Starting a discussion here since it doesn't seem like the Gitter is very active- is it possible for me to get a draft proposal reviewed before I upload the official one, with feedback maybe?

@ahartikainen @aloctavodia @amaloney @OriolAbril

**Comment by aloctavodia** [comment]
Hi @imperorrp, thanks for reaching out. Yes, we can provide feedback. We suggest you work on the Plotting Restructuring project. The reason is that we have just decided to rework the arviz-dashboard base code, and that will make the work of a student very difficult. What do you think?

**Reply by imperorrp** [reply]
Sure, I can do that- I also noticed that apart from the links provided with the idea concept for Plotting Restructuring, https://github.com/arviz-devs/arviz-plots also exists- does this have something to do with the plotting restructuring work going on as well?

**Reply by OriolAbril** [reply]
Yes, you'll contribute to the arviz-plots repository. We have started moving from the prototype discussion to an alpha version library.

The plotting backend project stays the same with the only difference of contributing to arviz_plots.backend.

The data organization will have to be modified a bit. We have decided to move forward with PlotCollection only, some docs and tests (still in a PR only) have already been written but both need significant work, and the API is starting to stabilize, so that part should probably be partially replaced with instead using PlotCollection class to implement "batteries-included" plots (how and how much we can discuss depending on your preferences too, even adapt somewhat as we go). That is, use this class to rewrite plots from current arviz; so far plot_forest and plot_dist are nearly finished and should serve as examples, plot_trace and plot_trace_dist still need some more work but will serve as templates too by the time gsoc coding period starts.

**Reply by imperorrp** [reply]
Thank you for the clarification! So xrtist.backend was just the previous iteration, and the current work in that regard is on arviz_plots.backend if I got that right? Also, what sort of functions exactly have to be implemented here- at least the ones in the existing base Arviz functionality? 

And I noticed only PlotCollection was mentioned in those docs you mentioned, yes. With the "batteries-included" plots, are they essentially aesthetic improvements to the preexisting plot functionalities? So work would involve adding these kinds of improvements to all the plots (apart from plot_forest/plot_dist/plot_trace/plot_trace_dist which would probably be done by then like you said), right?

**Reply by aloctavodia** [reply]
You are right active work is on arviz-plots repo.


We are planning to reinplement most plots currently on arviz. Some of them are going to be removed and some will be split into two different plots. This is partially reflected in the list of open issues on arviz-plots. The improvements are not only aesthetics, the new API will be more flexible and more compact. On the devs side plots will be easier to write, maintain and extend.

**Reply by imperorrp** [reply]
Okay, that makes sense. As for the backend, I suppose the new functions to be added there would be those that support or assist in the plotting of the new plots to be implemented (via the PlotCollection class), right?

**Reply by imperorrp** [reply]
Hello @aloctavodia @OriolAbril sorry for the delay! Here's [my draft proposal](https://docs.google.com/document/d/1tOFZDvHVTuXeNmPnYZfs7kwRPOY54kHvQNMOr5P3k-k/edit?usp=sharing). I've estimated a rough figure of 8 plots to refactor into ArviZ-Plot- what are your thoughts on this? Let me know if you have any other recommendations or thoughts as well for the proposal

**Reply by OriolAbril** [reply]
Proposal looks good, thanks for sharing. I am currently updating issues to hopefully make things easier for you to get onboard. We haven't had much community involvement yet for arviz-plots so we haven't been very good at keeping issues to date and the like. Here are a couple minor points that you might want to update in the proposal:

* support for multiple models (that is, have plotting functions accept not only DataTree objects but also dictionaries of DataTree objects) was added with plot_forest in https://github.com/arviz-devs/arviz-plots/pull/34. So that limitation for plot_dens is technically gone.
* we decided to not follow up with the plot_trace_dens name but instead use plot_trace_dist (you'll see it matches the description in plot_trace_dens issue), likewise, plot_rank_dens will be plot_rank_dist.

Now on to the actual proposal comments. I agree that leaving the exact list of plots to be added to be finalized at the beginning of the project is a good idea, but I would like to know what are/would be your preferences, mostly because that would help us prepare start the project on a good place. Some examples of what I mean:

* you have plot_ppc on the list, with a note on groupby behaviour. The groupby behaviour would require changes to PlotCollection itself as it would be a new largish-feature. Is plot_ppc high on your preference list? Is the grouby behaviour something that would interest you, thus working directly on PlotCollection itself (something challenging in different ways) or would you rather focus on using it?
* You have mentioned adding tests as you go, and we already do have some tests, but as this is a library on early stages, we are also still seeing how is it best to test things efficiently and extensively. Would working on test design and infrastructure be interesting or would you rather be given test patterns and templates and use them on the new plots?
* After reading the proposal I am not sure how familiar you are with matplotlib, bokeh, xarray, pytest and sphinx, all of which are key to the project. Not knowing them already is not an issue, but knowing it can help set a sensible scope and again, focus on developer docs on the relevant areas or at least gathering resources.

None of these answers will make your proposal more or less likely to be accepted, they will help us narrow down the scope a bit and make sure you'll be able to execute the proposal during the summer (i.e. by prioritizing setting up testing infrastructure over new plots or documentation or vice-versa).

**Reply by imperorrp** [reply]
Thank you for the review! Apart from what you said, I just saw you made a suggestion about omitting visual elements in an issue too- so I suppose plot_dens can be omitted since its functionality could be gained through plot_dist with visual element omission (once a solution to that is found, of course)? 

I will make these updates to the proposal right away- also removing plot_trace_dens and replacing plot_rank_dens with plot_rank_dist. I'll also add a bit about learning the learning the dependencies required for this project as you mention too (of course I'd have to at some point to execute it)

About my preferences as you ask- 

- The challenge would be interesting, yes- I can work on the PlotCollection class to support it. What I could do, since a change of this sort would be large, is leave plot_ppc towards the end if selected, when I would likely have a better understanding of the PlotCollection class and how to approach this issue. If everything else is on time then that should allow for some time at the end to do this
- About the question on testing, I think I am leaning towards the latter- having test patterns and templates to use for the new plots and modifying them for each individual plot seems better. I am also not familiar with using pytest before, so although I'll definitely have to get more familiar with it, some structure seems good. 
- I've used matplotlib/bokeh before for some basic visuals, but I would likely have to get more familiar with them yes. With xarray/pytest/sphinx you are right but I do plan on getting familiar with them before the coding period starts.
